File: questions/0876-Middle-of-the-Linked-List/legends/lee215.py
# Each time, slow go 1 steps while fast go 2 steps.
# When fast arrives at the end, slow will arrive right in the middle.

# Python:

def middleNode(self, head):
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    return slow

File: questions/0876-Middle-of-the-Linked-List/legends/lee215.cpp
// Each time, slow go 1 steps while fast go 2 steps.
// When fast arrives at the end, slow will arrive right in the middle.

// C++:

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode *middleNode(ListNode *head)
{
    ListNode *slow = head, *fast = head;
    while (fast && fast->next)
        slow = slow->next, fast = fast->next->next;
    return slow;
}